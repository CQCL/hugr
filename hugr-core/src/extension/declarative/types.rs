//! Declarative type definitions.
//!
//! This module defines a YAML schema for defining types in a declarative way.
//!
//! See the [specification] and [`ExtensionSetDeclaration`] for more details.
//!
//! [specification]: https://github.com/CQCL/hugr/blob/main/specification/hugr.md#declarative-format
//! [`ExtensionSetDeclaration`]: super::ExtensionSetDeclaration

use std::sync::Weak;

use crate::Extension;
use crate::extension::{TypeDef, TypeDefBound};
use crate::types::type_param::{TermEnum, TypeParam};
use crate::types::{Term, TypeBound, TypeName};

use serde::{Deserialize, Serialize};

use super::{DeclarationContext, ExtensionDeclarationError};

/// A declarative type definition.
#[derive(Debug, Clone, Serialize, Deserialize, Hash, PartialEq, Eq)]
pub(super) struct TypeDeclaration {
    /// The name of the type.
    name: TypeName,
    /// A description for the type.
    #[serde(default)]
    #[serde(skip_serializing_if = "crate::utils::is_default")]
    description: String,
    /// The type bound describing what can be done to instances of this type.
    /// Options are `Eq`, `Copyable`, or `Any`.
    ///
    /// See [`TypeBound`] and [`TypeDefBound`].
    ///
    /// TODO: Derived bounds from the parameters (see [`TypeDefBound`]) are not yet supported.
    #[serde(default)]
    #[serde(skip_serializing_if = "crate::utils::is_default")]
    bound: TypeDefBoundDeclaration,
    /// A list of type parameters for this type.
    ///
    /// Each element in the list is a 2-element list, where the first element is
    /// the human-readable name of the type parameter, and the second element is
    /// the type id.
    #[serde(default)]
    #[serde(skip_serializing_if = "crate::utils::is_default")]
    params: Vec<TypeParamDeclaration>,
}

impl TypeDeclaration {
    /// Register this type in the given extension.
    ///
    /// Types in the definition will be resolved using the extensions in `scope`
    /// and the current extension.
    ///
    /// Requires a [`Weak`] reference to the extension defining the operation.
    /// This method is intended to be used inside the closure passed to [`Extension::new_arc`].
    pub fn register<'ext>(
        &self,
        ext: &'ext mut Extension,
        ctx: DeclarationContext<'_>,
        extension_ref: &Weak<Extension>,
    ) -> Result<&'ext TypeDef, ExtensionDeclarationError> {
        let params = self
            .params
            .iter()
            .map(|param| param.make_type_param(ext, ctx))
            .collect::<Result<Vec<TypeParam>, _>>()?;
        let type_def = ext.add_type(
            self.name.clone(),
            params,
            self.description.clone(),
            self.bound.into(),
            extension_ref,
        )?;
        Ok(type_def)
    }
}

/// A declarative `TypeBound` definition.
///
/// Equivalent to a [`TypeDefBound`]. Provides human-friendly serialization, using
/// the full names.
///
/// TODO: Support derived bounds
#[derive(
    Debug, Copy, Clone, Serialize, Deserialize, Hash, PartialEq, Eq, Default, derive_more::Display,
)]
enum TypeDefBoundDeclaration {
    /// The type can be copied in the program.
    Copyable,
    /// No bound on the type.
    #[default]
    Any,
}

impl From<TypeDefBoundDeclaration> for TypeDefBound {
    fn from(bound: TypeDefBoundDeclaration) -> Self {
        match bound {
            TypeDefBoundDeclaration::Copyable => Self::Explicit {
                bound: TypeBound::Copyable,
            },
            TypeDefBoundDeclaration::Any => Self::Explicit {
                bound: TypeBound::Any,
            },
        }
    }
}

/// A declarative type parameter definition.
///
/// Only supports [`TypeParam::String`]s for now.
#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
#[serde(untagged)]
enum TypeParamDeclaration {
    String,
    /// A 2-tuple containing a human-readable name and a type id.
    WithDescription(String),
}

impl TypeParamDeclaration {
    /// Create a [`TypeParam`] from this declaration.
    ///
    /// Resolves any type ids using both the current extension and any other in `scope`.
    ///
    /// TODO: Only non-parametric opaque types are supported for now.
    /// TODO: The parameter description is currently ignored.
    pub fn make_type_param(
        &self,
        _extension: &Extension,
        _ctx: DeclarationContext<'_>,
    ) -> Result<TypeParam, ExtensionDeclarationError> {
        Ok(Term::new(TermEnum::StringType))
    }
}
