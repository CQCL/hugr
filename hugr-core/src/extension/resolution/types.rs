//! Resolve weak links inside `CustomType`s in an optype's signature.

use std::sync::Arc;

use super::{ExtensionRegistry, ExtensionResolutionError};
use crate::ops::OpType;
use crate::types::type_row::TypeRowBase;
use crate::types::{CustomType, MaybeRV, Signature, SumType, TypeBase, TypeEnum};
use crate::Node;

/// Replace the dangling extension pointer in the [`CustomType`]s inside a
/// signature with a valid pointer to the extension in the `extensions`
/// registry.
///
/// When a pointer is replaced, the extension is added to the
/// `used_extensions` registry and the new type definition is returned.
///
/// This is a helper function used right after deserializing a Hugr.
pub fn update_op_types_extensions(
    node: Node,
    op: &mut OpType,
    extensions: &ExtensionRegistry,
    used_extensions: &mut ExtensionRegistry,
) -> Result<(), ExtensionResolutionError> {
    match op {
        OpType::ExtensionOp(ext) => {
            update_signature_exts(node, ext.signature_mut(), extensions, used_extensions)?
        }
        OpType::FuncDefn(f) => {
            update_signature_exts(node, f.signature.body_mut(), extensions, used_extensions)?
        }
        OpType::FuncDecl(f) => {
            update_signature_exts(node, f.signature.body_mut(), extensions, used_extensions)?
        }
        OpType::Const(_c) => {
            // TODO: Is it OK to assume that `Value::get_type` returns a well-resolved value?
        }
        OpType::Input(inp) => {
            update_type_row_exts(node, &mut inp.types, extensions, used_extensions)?
        }
        OpType::Output(out) => {
            update_type_row_exts(node, &mut out.types, extensions, used_extensions)?
        }
        OpType::Call(c) => {
            update_signature_exts(node, c.func_sig.body_mut(), extensions, used_extensions)?;
            update_signature_exts(node, &mut c.instantiation, extensions, used_extensions)?;
        }
        OpType::CallIndirect(c) => {
            update_signature_exts(node, &mut c.signature, extensions, used_extensions)?
        }
        OpType::LoadConstant(lc) => {
            update_type_exts(node, &mut lc.datatype, extensions, used_extensions)?
        }
        OpType::LoadFunction(lf) => {
            update_signature_exts(node, lf.func_sig.body_mut(), extensions, used_extensions)?;
            update_signature_exts(node, &mut lf.signature, extensions, used_extensions)?;
        }
        OpType::DFG(dfg) => {
            update_signature_exts(node, &mut dfg.signature, extensions, used_extensions)?
        }
        OpType::OpaqueOp(op) => {
            update_signature_exts(node, op.signature_mut(), extensions, used_extensions)?
        }
        OpType::Tag(t) => {
            for variant in t.variants.iter_mut() {
                update_type_row_exts(node, variant, extensions, used_extensions)?
            }
        }
        OpType::DataflowBlock(db) => {
            update_type_row_exts(node, &mut db.inputs, extensions, used_extensions)?;
            update_type_row_exts(node, &mut db.other_outputs, extensions, used_extensions)?;
            for row in db.sum_rows.iter_mut() {
                update_type_row_exts(node, row, extensions, used_extensions)?;
            }
        }
        OpType::ExitBlock(e) => {
            update_type_row_exts(node, &mut e.cfg_outputs, extensions, used_extensions)?;
        }
        OpType::TailLoop(tl) => {
            update_type_row_exts(node, &mut tl.just_inputs, extensions, used_extensions)?;
            update_type_row_exts(node, &mut tl.just_outputs, extensions, used_extensions)?;
            update_type_row_exts(node, &mut tl.rest, extensions, used_extensions)?;
        }
        OpType::CFG(cfg) => {
            update_signature_exts(node, &mut cfg.signature, extensions, used_extensions)?;
        }
        OpType::Conditional(cond) => {
            for row in cond.sum_rows.iter_mut() {
                update_type_row_exts(node, row, extensions, used_extensions)?;
            }
            update_type_row_exts(node, &mut cond.other_inputs, extensions, used_extensions)?;
            update_type_row_exts(node, &mut cond.outputs, extensions, used_extensions)?;
        }
        OpType::Case(case) => {
            update_signature_exts(node, &mut case.signature, extensions, used_extensions)?;
        }
        // Ignore optypes that do not store a signature.
        _ => {}
    }
    Ok(())
}

/// Update all weak Extension pointers in the [`CustomType`]s inside a signature.
///
/// Adds the extensions used in the signature to the `used_extensions` registry.
fn update_signature_exts(
    node: Node,
    signature: &mut Signature,
    extensions: &ExtensionRegistry,
    used_extensions: &mut ExtensionRegistry,
) -> Result<(), ExtensionResolutionError> {
    update_type_row_exts(node, &mut signature.input, extensions, used_extensions)?;
    update_type_row_exts(node, &mut signature.output, extensions, used_extensions)?;
    Ok(())
}

/// Update all weak Extension pointers in the [`CustomType`]s inside a type row.
///
/// Adds the extensions used in the row to the `used_extensions` registry.
fn update_type_row_exts<RV: MaybeRV>(
    node: Node,
    row: &mut TypeRowBase<RV>,
    extensions: &ExtensionRegistry,
    used_extensions: &mut ExtensionRegistry,
) -> Result<(), ExtensionResolutionError> {
    for ty in row.iter_mut() {
        update_type_exts(node, ty, extensions, used_extensions)?;
    }
    Ok(())
}

/// Update all weak Extension pointers in the [`CustomType`]s inside a type.
///
/// Adds the extensions used in the type to the `used_extensions` registry.
fn update_type_exts<RV: MaybeRV>(
    node: Node,
    typ: &mut TypeBase<RV>,
    extensions: &ExtensionRegistry,
    used_extensions: &mut ExtensionRegistry,
) -> Result<(), ExtensionResolutionError> {
    match typ.as_type_enum_mut() {
        TypeEnum::Extension(custom) => {
            let ext_id = custom.extension();
            let ext = extensions.get(ext_id).ok_or_else(|| {
                ExtensionResolutionError::missing_type_extension(
                    node,
                    custom.name(),
                    ext_id,
                    extensions,
                )
            })?;

            // Add the extension to the used extensions registry,
            // and update the CustomType with the valid pointer.
            used_extensions.register_updated_ref(ext);
            *custom = CustomType::new(
                custom.name().clone(),
                custom.args(),
                custom.extension().clone(),
                custom.bound(),
                &Arc::downgrade(ext),
            );
        }
        TypeEnum::Function(f) => {
            update_type_row_exts(node, &mut f.input, extensions, used_extensions)?;
            update_type_row_exts(node, &mut f.output, extensions, used_extensions)?;
        }
        TypeEnum::Sum(SumType::General { rows }) => {
            for row in rows.iter_mut() {
                update_type_row_exts(node, row, extensions, used_extensions)?;
            }
        }
        _ => {}
    }
    Ok(())
}
