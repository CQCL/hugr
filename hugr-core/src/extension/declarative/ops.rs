//! Declarative operation definitions.
//!
//! This module defines a YAML schema for defining operations in a declarative way.
//!
//! See the [specification] and [`ExtensionSetDeclaration`] for more details.
//!
//! [specification]: https://github.com/CQCL/hugr/blob/main/specification/hugr.md#declarative-format
//! [`ExtensionSetDeclaration`]: super::ExtensionSetDeclaration

use std::collections::HashMap;

use serde::{Deserialize, Serialize};
use smol_str::SmolStr;

use crate::extension::{OpDef, SignatureFunc};
use crate::ops::OpName;
use crate::types::type_param::TypeParam;
use crate::Extension;

use super::signature::SignatureDeclaration;
use super::{DeclarationContext, ExtensionDeclarationError};

/// A declarative operation definition.
///
/// TODO: The "Lowering" attribute is not yet supported.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub(super) struct OperationDeclaration {
    /// The identifier the operation.
    name: OpName,
    /// A description for the operation.
    #[serde(default)]
    #[serde(skip_serializing_if = "crate::utils::is_default")]
    description: String,
    /// The signature of the operation.
    #[serde(default)]
    #[serde(skip_serializing_if = "crate::utils::is_default")]
    signature: Option<SignatureDeclaration>,
    /// A set of per-node parameters required to instantiate this operation.
    #[serde(default)]
    #[serde(skip_serializing_if = "crate::utils::is_default")]
    params: HashMap<SmolStr, ParamDeclaration>,
    /// An extra set of data associated to the operation.
    ///
    /// This data is kept in the Hugr, and may be accessed by the relevant runtime.
    #[serde(default)]
    #[serde(skip_serializing_if = "crate::utils::is_default")]
    misc: HashMap<String, serde_yaml::Value>,
    /// A pre-compiled lowering routine.
    ///
    /// This is not yet supported, and will raise an error if present.
    #[serde(default)]
    #[serde(skip_serializing_if = "crate::utils::is_default")]
    lowering: Option<String>,
}

impl OperationDeclaration {
    /// Register this operation in the given extension.
    pub fn register<'ext>(
        &self,
        ext: &'ext mut Extension,
        ctx: DeclarationContext<'_>,
    ) -> Result<&'ext mut OpDef, ExtensionDeclarationError> {
        // We currently only support explicit signatures.
        //
        // TODO: Support missing signatures?
        let Some(signature) = &self.signature else {
            return Err(ExtensionDeclarationError::MissingSignature {
                ext: ext.name().clone(),
                op: self.name.clone(),
            });
        };

        // We currently do not support parametric operations.
        if !self.params.is_empty() {
            return Err(ExtensionDeclarationError::ParametricOperation {
                ext: ext.name().clone(),
                op: self.name.clone(),
            });
        }
        let params: Vec<TypeParam> = vec![];

        if self.lowering.is_some() {
            return Err(ExtensionDeclarationError::LoweringNotSupported {
                ext: ext.name().clone(),
                op: self.name.clone(),
            });
        }

        let signature_func: SignatureFunc = signature.make_signature(ext, ctx, &params)?;

        let op_def = ext.add_op(self.name.clone(), self.description.clone(), signature_func)?;

        for (k, v) in &self.misc {
            op_def.add_misc(k, v.clone());
        }

        Ok(op_def)
    }
}

/// The type of a per-node operation parameter required to instantiate an operation.
///
/// TODO: The value should be decoded as a [`TypeParam`].
/// Valid options include:
///
/// - `USize`
/// - `Type`
///
/// [`TypeParam`]: crate::types::type_param::TypeParam
#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
pub(super) struct ParamDeclaration(
    /// TODO: Store a [`TypeParam`], and implement custom parsers.
    ///
    /// [`TypeParam`]: crate::types::type_param::TypeParam
    String,
);
