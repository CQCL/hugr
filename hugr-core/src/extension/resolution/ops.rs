//! Resolve `OpaqueOp`s into `ExtensionOp`s and return an operation's required
//! extension.
//!
//! Contains both mutable ([`resolve_op_extensions`]) and immutable
//! ([`collect_operation_extension`]) methods to resolve operations and collect
//! the required extensions respectively.

use std::sync::Arc;

use super::{Extension, ExtensionCollectionError, ExtensionResolutionError};
use crate::Node;
use crate::extension::ExtensionRegistry;
use crate::ops::custom::OpaqueOpError;
use crate::ops::{DataflowOpTrait, ExtensionOp, NamedOp, OpType};

/// Returns the extension in the registry required by the operation.
///
/// If the operation does not require an extension, returns `None`.
///
/// [`ExtensionOp`]s store a [`Weak`] reference to their extension, which can be
/// invalidated if the original `Arc<Extension>` is dropped. On such cases, we
/// return an error with the missing extension names.
///
/// # Parameters
///
/// - `node`: The node where the operation is located, if available. This is
///   used to provide context in the error message.
/// - `op`: The operation to collect the extensions from.
pub(crate) fn collect_op_extension(
    node: Option<Node>,
    op: &OpType,
) -> Result<Option<Arc<Extension>>, ExtensionCollectionError> {
    let OpType::ExtensionOp(ext_op) = op else {
        // TODO: Extract the extension when the operation is a `Const`.
        // https://github.com/CQCL/hugr/issues/1742
        return Ok(None);
    };
    let ext = ext_op.def().extension();
    match ext.upgrade() {
        Some(e) => Ok(Some(e)),
        None => Err(ExtensionCollectionError::dropped_op_extension(
            node,
            op,
            [ext_op.def().extension_id().clone()],
        )),
    }
}

/// Compute the required extension for an operation.
///
/// If the op is a [`OpType::OpaqueOp`], replace it with a resolved
/// [`OpType::ExtensionOp`] by looking searching for the operation in the
/// extension registries.
///
/// If `op` was an opaque or extension operation, the result contains the
/// extension reference that should be added to the hugr's extension registry.
///
/// # Errors
///
/// If the serialized opaque resolves to a definition that conflicts with what
/// was serialized. Or if the operation is not found in the registry.
pub(crate) fn resolve_op_extensions<'e>(
    node: Node,
    op: &mut OpType,
    extensions: &'e ExtensionRegistry,
) -> Result<Option<&'e Arc<Extension>>, ExtensionResolutionError> {
    let extension = operation_extension(node, op, extensions)?;

    let OpType::OpaqueOp(opaque) = op else {
        return Ok(extension);
    };

    // Fail if the Extension is not in the registry, or if the Extension was
    // found but did not have the expected operation.
    let extension = extension.expect("OpaqueOp should have an extension");
    let Some(def) = extension.get_op(opaque.unqualified_id()) else {
        return Err(OpaqueOpError::OpNotFoundInExtension {
            node,
            op: opaque.name().clone(),
            extension: extension.name().clone(),
            available_ops: extension
                .operations()
                .map(|(name, _)| name.clone())
                .collect(),
        }
        .into());
    };

    let ext_op = ExtensionOp::new_with_cached(def.clone(), opaque.args().to_vec(), opaque)
        .map_err(|e| OpaqueOpError::SignatureError {
            node,
            name: opaque.name().clone(),
            cause: e,
        })?;

    if opaque.signature().io() != ext_op.signature().io() {
        return Err(OpaqueOpError::SignatureMismatch {
            node,
            extension: opaque.extension().clone(),
            op: def.name().clone(),
            computed: ext_op.signature().into_owned(),
            stored: opaque.signature().into_owned(),
        }
        .into());
    }

    // Replace the opaque operation with the resolved extension operation.
    *op = ext_op.into();

    Ok(Some(extension))
}

/// Returns the extension in the registry required by the operation.
///
/// If the operation does not require an extension, returns `None`.
fn operation_extension<'e>(
    node: Node,
    op: &OpType,
    extensions: &'e ExtensionRegistry,
) -> Result<Option<&'e Arc<Extension>>, ExtensionResolutionError> {
    let Some(ext) = op.extension_id() else {
        return Ok(None);
    };
    match extensions.get(ext) {
        Some(e) => Ok(Some(e)),
        None => Err(ExtensionResolutionError::missing_op_extension(
            Some(node),
            op,
            ext,
            extensions,
        )),
    }
}
