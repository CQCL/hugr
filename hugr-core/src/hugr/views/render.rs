//! Helper methods to compute the node/edge/port style when rendering a HUGR
//! into dot or mermaid format.

use portgraph::render::{EdgeStyle, NodeStyle, PortStyle};
use portgraph::{LinkView, NodeIndex, PortIndex, PortView};

use crate::ops::NamedOp;
use crate::types::EdgeKind;
use crate::HugrView;

/// Configuration for rendering a HUGR graph.
#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
#[non_exhaustive]
pub struct RenderConfig {
    /// Show the node index in the graph nodes.
    pub node_indices: bool,
    /// Show port offsets in the graph edges.
    pub port_offsets_in_edges: bool,
    /// Show type labels on edges.
    pub type_labels_in_edges: bool,
}

impl Default for RenderConfig {
    fn default() -> Self {
        Self {
            node_indices: true,
            port_offsets_in_edges: true,
            type_labels_in_edges: true,
        }
    }
}

/// Formatter method to compute a node style.
pub(super) fn node_style<H: HugrView + ?Sized>(
    h: &H,
    config: RenderConfig,
) -> Box<dyn FnMut(NodeIndex) -> NodeStyle + '_> {
    if config.node_indices {
        Box::new(move |n| {
            NodeStyle::Box(format!(
                "({ni}) {name}",
                ni = n.index(),
                name = h.get_optype(n.into()).name()
            ))
        })
    } else {
        Box::new(move |n| NodeStyle::Box(h.get_optype(n.into()).name().to_string()))
    }
}

/// Formatter method to compute a port style.
pub(super) fn port_style<H: HugrView + ?Sized>(
    h: &H,
    _config: RenderConfig,
) -> Box<dyn FnMut(PortIndex) -> PortStyle + '_> {
    let graph = h.portgraph();
    Box::new(move |port| {
        let node = graph.port_node(port).unwrap();
        let optype = h.get_optype(node.into());
        let offset = graph.port_offset(port).unwrap();
        match optype.port_kind(offset).unwrap() {
            EdgeKind::Function(pf) => PortStyle::new(html_escape::encode_text(&format!("{}", pf))),
            EdgeKind::Const(ty) | EdgeKind::Value(ty) => {
                PortStyle::new(html_escape::encode_text(&format!("{}", ty)))
            }
            EdgeKind::StateOrder => match graph.port_links(port).count() > 0 {
                true => PortStyle::text("", false),
                false => PortStyle::Hidden,
            },
            _ => PortStyle::text("", true),
        }
    })
}

/// Formatter method to compute an edge style.
#[allow(clippy::type_complexity)]
pub(super) fn edge_style<H: HugrView + ?Sized>(
    h: &H,
    config: RenderConfig,
) -> Box<
    dyn FnMut(
            <H::Portgraph<'_> as LinkView>::LinkEndpoint,
            <H::Portgraph<'_> as LinkView>::LinkEndpoint,
        ) -> EdgeStyle
        + '_,
> {
    let graph = h.portgraph();
    Box::new(move |src, tgt| {
        let src_node = graph.port_node(src).unwrap();
        let src_optype = h.get_optype(src_node.into());
        let src_offset = graph.port_offset(src).unwrap();
        let tgt_offset = graph.port_offset(tgt).unwrap();

        let port_kind = src_optype.port_kind(src_offset).unwrap();

        // StateOrder edges: Dotted line.
        // Control flow edges: Dashed line.
        // Static and Value edges: Solid line with label.
        let style = match port_kind {
            EdgeKind::StateOrder => EdgeStyle::Dotted,
            EdgeKind::ControlFlow => EdgeStyle::Dashed,
            EdgeKind::Const(_) | EdgeKind::Function(_) | EdgeKind::Value(_) => EdgeStyle::Solid,
        };

        // Compute the label for the edge, given the setting flags.
        fn type_label(e: EdgeKind) -> Option<String> {
            match e {
                EdgeKind::Const(ty) | EdgeKind::Value(ty) => Some(format!("{}", ty)),
                EdgeKind::Function(pf) => Some(format!("{}", pf)),
                _ => None,
            }
        }
        //
        // Only static and value edges have types to display.
        let label = match (
            config.port_offsets_in_edges,
            type_label(port_kind).filter(|_| config.type_labels_in_edges),
        ) {
            (true, Some(ty)) => {
                format!("{}:{}\n{ty}", src_offset.index(), tgt_offset.index())
            }
            (true, _) => format!("{}:{}", src_offset.index(), tgt_offset.index()),
            (false, Some(ty)) => ty.to_string(),
            _ => return style,
        };
        style.with_label(label)
    })
}
