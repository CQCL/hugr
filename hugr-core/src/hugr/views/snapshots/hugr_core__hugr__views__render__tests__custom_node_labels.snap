---
source: hugr-core/src/hugr/views/render.rs
expression: h.mermaid_string_with_formatter(config)
---
graph LR
    subgraph 0 ["(node_0) Module"]
        direction LR
        subgraph 1 ["(node_1) FuncDefn: #quot;main#quot;"]
            direction LR
            2["(node_2) Input"]
            3["(node_3) Output"]
            subgraph 4 ["(node_4) [**DFG**]"]
                direction LR
                style 4 stroke:#832561,stroke-width:3px
                5["(node_5) Input"]
                6["(node_6) Output"]
                5--"0:0<br>Bool"-->6
            end
            2--"0:0<br>Bool"-->4
            4--"0:0<br>Bool"-->3
        end
    end
